### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by Dmytrit**
[DEMO LINK](https://Dmytrit.github.io/react_dynamic-list-of-todos/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,14 +1,42 @@
...
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
 
 export const App: React.FC = () => {
```
Comment (Prince3378) [inline src/App.tsx:13]: Виглядає гарно! Але треба підправити пошук, він працює тільки на вкладці All

**Review by Prince3378 — APPROVED** [review]

File: src/App.tsx [hunk 2]
```diff
@@ -1,14 +1,42 @@
...
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterValue, setFilterValue] = useState('all');
```
Comment (jstmpelowycz) [inline src/App.tsx:15]: use enum
```suggestion
  const [filterValue, setFilterValue] = useState(FilterType.All);
```

File: src/App.tsx [hunk 3]
```diff
@@ -1,14 +1,42 @@
...
+  const [query, setQuery] = useState('');
+
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+
+  useEffect(() => {
+    getTodos()
+      .then(todosFromServer => setTodos([...todosFromServer]));
```
Comment (jstmpelowycz) [inline src/App.tsx:22]:
```suggestion
      .then(todosFromServer => setTodos(todosFromServer));
```

File: src/App.tsx [hunk 4]
```diff
@@ -1,14 +1,42 @@
...
+  }, []);
+
+  const visibleTodos = todos.filter(todo => {
+    if (filterValue === 'active') {
+      return !todo.completed;
+    }
+
+    if (filterValue === 'completed') {
+      return todo.completed;
+    }
```
Comment (jstmpelowycz) [inline src/App.tsx:32]: you'd rather use switch-case here

File: src/components/TodoList/TodoList.tsx [hunk 5]
```diff
@@ -1,100 +1,85 @@
...
-            <i className="fas fa-check" />
-          </span>
-        </th>
-        <th>Title</th>
-        <th> </th>
-      </tr>
-    </thead>
+type Props = {
+  todos: Todo[],
+  selectedTodoId?: number | null,
```
Comment (jstmpelowycz) [inline src/components/TodoList/TodoList.tsx:7]:
```suggestion
  selectedTodoId?: number | null,
```
if property is indicated as optional, you don't need it to be `null`, apply either

File: src/components/TodoModal/TodoModal.tsx [hunk 6]
```diff
@@ -1,12 +1,30 @@
...
+
+export const TodoModal: React.FC<Props> = ({
+  selectedTodo,
+  cleanSelectedTodo,
+}) => {
+  const [user, setUser] = useState<User | null>(null);
+
+  useEffect(() => {
+    getUser(selectedTodo.userId)
+      .then(curUser => setUser(curUser));
```
Comment (jstmpelowycz) [inline src/components/TodoModal/TodoModal.tsx:20]:
```suggestion
      .then(setUser);
```

**Review by jstmpelowycz — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 7]
```diff
@@ -1,14 +1,65 @@
...
+
+        break;
+
+      case FilterType.Active:
+        setVisibleTodos([...todos].filter(todo => !todo.completed));
+
+        break;
+
+      case FilterType.Completed:
+        setVisibleTodos([...todos].filter(todo => todo.completed));
```
Comment (jstmpelowycz) [inline src/App.tsx:47]: does `filter` mutate your array?

Comment (jstmpelowycz) [reply src/App.tsx:47]: I believe you know about it

File: src/components/TodoModal/TodoModal.tsx [hunk 8]
```diff
@@ -15,30 +33,38 @@ export const TodoModal: React.FC = () => {
...
-
+              {selectedTodo.completed ? (
+                <strong className="has-text-success">
+                  Done
+                </strong>
+              ) : (
+                <strong className="has-text-danger">
+                  Planned
+                </strong>
+              )}
```
Comment (jstmpelowycz) [inline src/components/TodoModal/TodoModal.tsx:64]: apply ternaries to content, not tags

File: src/components/TodoList/TodoList.tsx [hunk 9]
```diff
@@ -1,100 +1,85 @@
...
-          </span>
-        </th>
-        <th>Title</th>
-        <th> </th>
-      </tr>
-    </thead>
+type Props = {
+  todos: Todo[],
+  selectedTodoId?: number,
+  setSelectedTodo: (todo: Todo) => void
```
Comment (jstmpelowycz) [inline src/components/TodoList/TodoList.tsx:8]:
```suggestion
  onSetSelectedTodo: (todo: Todo) => void
```

**Review by jstmpelowycz — APPROVED** [review]

**Pull request closed without merging**